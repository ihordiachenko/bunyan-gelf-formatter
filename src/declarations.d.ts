declare module 'safe-flat' {
  function flatten(obj: any): any;
}
