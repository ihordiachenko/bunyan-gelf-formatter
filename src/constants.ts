/**
 * Syslog severity levels according to RFC-5424
 *
 * @link https://tools.ietf.org/html/rfc5424#section-6.2.1
 */
export enum SYSLOG_LEVELS {
  /**
   * System is unusable	A panic condition
   */
  EMERGENCY = 0,

  /**
   * Action must be taken immediately
   * A condition that should be corrected immediately, such as a corrupted system database.
   */
  ALERT = 1,

  /**
   * Critical conditions. Hard device errors.
   */
  CRITICAL = 2,

  /**
   * Error conditions.
   */
  ERROR = 3,

  /**
   * Warning conditions.
   */
  WARNING = 4,

  /**
   * Normal but significant conditions.
   * Conditions that are not error conditions, but that may require special handling.
   */
  NOTICE = 5,

  /**
   * Informational messages.
   */
  INFO = 6,

  /**
   * Debug messages.
   */
  DEBUG = 7,
}
