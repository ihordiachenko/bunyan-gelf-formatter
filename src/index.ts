import stream from 'stream';
import Logger from 'bunyan';
import flatten from 'flat';
import ErrorStackParser from 'error-stack-parser';
import { SYSLOG_LEVELS } from './constants';

export const GELF_VERSION = '1.1';
export { SYSLOG_LEVELS } from './constants';

const BUNYAN_TO_SYSYLOG_LEVELS = new Map([
  [Logger.FATAL, SYSLOG_LEVELS.CRITICAL],
  [Logger.ERROR, SYSLOG_LEVELS.ERROR],
  [Logger.WARN, SYSLOG_LEVELS.WARNING],
  [Logger.INFO, SYSLOG_LEVELS.INFO],
  [Logger.TRACE, SYSLOG_LEVELS.DEBUG],
  [Logger.DEBUG, SYSLOG_LEVELS.DEBUG],
]);

/**
 * Formats bunyan logs according to GELF payload spec.
 *
 * @link https://docs.graylog.org/en/3.2/pages/gelf.html?highlight=levels#gelf-payload-specification
 */
export class GelfStream extends stream.Transform {
  private _raw: boolean;

  constructor(options: stream.TransformOptions = {}, raw: boolean = false) {
    options.readableObjectMode = true;
    options.writableObjectMode = true;
    super(options);

    this._raw = raw;
  }

  _transform(
    chunk: any,
    _encoding: BufferEncoding,
    callback: stream.TransformCallback
  ): void {
    try {
      const gelfObject = this._bunyanToGelf(chunk);

      if (this._raw) {
        callback(null, gelfObject);
      } else {
        const serialized = JSON.stringify(gelfObject) + '\n';
        callback(null, serialized);
      }
    } catch (err) {
      callback(err);
    }
  }

  _bunyanToGelf(bunyanMessage: any) {
    const gelfMessage: any = {
      version: GELF_VERSION,
      host: bunyanMessage.hostname,
      timestamp: Math.floor(Number(new Date(bunyanMessage.time)) / 1000),
      short_message: bunyanMessage.msg,
      level: this._convertLevel(bunyanMessage.level),
      facility: bunyanMessage.name,
      ...omit(flatten(bunyanMessage), [
        'hostname',
        'time',
        'msg',
        'name',
        'level',
        'v',
        'err',
      ]),
    };

    if (bunyanMessage.err && bunyanMessage.err.stack) {
      gelfMessage.full_message = bunyanMessage.err.stack;
      const err = bunyanMessage.err;
      const { fileName, lineNumber } = ErrorStackParser.parse(err)[0];

      if (fileName) {
        gelfMessage.file = fileName;
      }
      if (lineNumber) {
        gelfMessage.line = Number(lineNumber);
      }
    }

    return gelfMessage;
  }

  _convertLevel(level: number) {
    return BUNYAN_TO_SYSYLOG_LEVELS.get(level) ?? SYSLOG_LEVELS.WARNING;
  }
}

function omit(obj: any, keys: string[]): any {
  const omitted = new Set<string>(keys);

  return Object.keys(obj)
    .filter(key => !omitted.has(key))
    .reduce((acc: any, key: string): any => {
      acc[key] = obj[key];
      return acc;
    }, {});
}
