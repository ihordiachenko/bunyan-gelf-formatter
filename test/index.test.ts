import { GelfFormatStream } from '../src';

describe('blah', () => {
  let stream: GelfFormatStream;

  describe('_transform', () => {
    beforeEach(() => {
      stream = new GelfFormatStream({}, true);
    });

    it('should return an object in raw mode', () => {});

    it('should return stringified json in normal mode', () => {});
  });
});
