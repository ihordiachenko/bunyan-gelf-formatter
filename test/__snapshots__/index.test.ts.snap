// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bunyan-gelf-formatter normal mode should write stringified json: stringified 1`] = `
"{\\"version\\":\\"1.1\\",\\"host\\":\\"jest-host\\",\\"timestamp\\":1581638400,\\"short_message\\":\\"oh my!\\",\\"level\\":3,\\"facility\\":\\"test-logger\\",\\"pid\\":18768,\\"full_message\\":\\"Error: The roof is on fire\\\\n    at Object.<anonymous> (/Users/user/repos/bunyan-gelf-formatter/test/index.test.ts:8:15)\\\\n    at Runtime._execModule (/Users/user/repos/bunyan-gelf-formatter/node_modules/jest-runtime/build/index.js:1205:24)\\\\n    at Runtime._loadModule (/Users/user/repos/bunyan-gelf-formatter/node_modules/jest-runtime/build/index.js:805:12)\\\\n    at Runtime.requireModule (/Users/user/repos/bunyan-gelf-formatter/node_modules/jest-runtime/build/index.js:662:10)\\\\n    at jasmine2 (/Users/user/repos/bunyan-gelf-formatter/node_modules/jest-jasmine2/build/index.js:228:13)\\\\n    at runTestInternal (/Users/user/repos/bunyan-gelf-formatter/node_modules/jest-runner/build/runTest.js:385:22)\\\\n    at runTest (/Users/user/repos/bunyan-gelf-formatter/node_modules/jest-runner/build/runTest.js:465:34)\\",\\"file\\":\\"/Users/user/repos/bunyan-gelf-formatter/test/index.test.ts\\",\\"line\\":8}
"
`;
